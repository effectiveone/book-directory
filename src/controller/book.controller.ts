import {
  Controller,
  Get,
  Post,
  Body,
  Param,
  Put,
  Delete,
} from '@nestjs/common';
import { BookService } from '../services/book.service';
import { Book } from '../model/book.model';

@Controller('books')
export class BookController {
  constructor(private bookService: BookService) {}

  @Post()
  async create(@Body() book: Book): Promise<Book> {
    return this.bookService.create(book);
  }

  @Get()
  async findAll(): Promise<Book[]> {
    return this.bookService.findAll();
  }

  @Get(':id')
  async findOne(@Param('id') id: string): Promise<Book> {
    return this.bookService.findOne(id);
  }

  @Put(':id')
  async update(@Param('id') id: string, @Body() book: Book): Promise<Book> {
    return this.bookService.update(id, book);
  }

  @Delete(':id')
  async delete(@Param('id') id: string): Promise<any> {
    return this.bookService.delete(id);
  }
}
