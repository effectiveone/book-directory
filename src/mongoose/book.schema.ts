import { Schema, Document } from 'mongoose';

export interface Book extends Document {
  id: string;
  title: string;
  author: string;
  yearPublished: number;
}

export const BookSchema = new Schema({
  id: { type: String, required: true },
  title: { type: String, required: true },
  author: { type: String, required: true },
  yearPublished: { type: Number, required: true },
});

export default BookSchema;
