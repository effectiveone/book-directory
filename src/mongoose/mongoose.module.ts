import { Module } from '@nestjs/common';
import { MongooseModule as NestMongooseModule } from '@nestjs/mongoose';
import { BookSchema } from './book.schema';

@Module({
  imports: [
    NestMongooseModule.forRoot('mongodb://localhost/book-directory', {
      connectionFactory: (connection) => {
        connection.plugin(require('mongoose-autopopulate'));
        return connection;
      },
      useNewUrlParser: true,
      useUnifiedTopology: true,
    }),
    NestMongooseModule.forFeature([{ name: 'Book', schema: BookSchema }]),
  ],
  exports: [NestMongooseModule],
})
export class MongooseModule {}
