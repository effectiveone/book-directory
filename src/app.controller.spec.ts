import { Test, TestingModule } from '@nestjs/testing';
import { INestApplication } from '@nestjs/common';
import { getModelToken } from '@nestjs/mongoose';
import * as request from 'supertest';
import { Book } from './interfaces/book.interface';
import { BookService } from './services/book.service';
import { BookController } from './controller/book.controller';

describe('AppController (e2e)', () => {
  let app: INestApplication;
  let bookModel;

  beforeAll(async () => {
    const moduleFixture: TestingModule = await Test.createTestingModule({
      controllers: [BookController],
      providers: [
        BookService,
        {
          provide: getModelToken('Book'),
          useValue: {
            find: () => ({
              exec: () => [
                { id: 'book1', title: 'Book 1', author: 'Author 1' },
                { id: 'book2', title: 'Book 2', author: 'Author 2' },
                { id: 'book3', title: 'Book 3', author: 'Author 3' },
              ],
            }),
          },
        },
      ],
    }).compile();

    app = moduleFixture.createNestApplication();
    await app.init();

    bookModel = moduleFixture.get(getModelToken('Book'));
  });

  describe('/books (GET)', () => {
    it('should return an array of books', () => {
      return request(app.getHttpServer())
        .get('/books')
        .expect(200)
        .expect((response) => {
          const books: Book[] = response.body;
          expect(books.length).toEqual(3);
          expect(books[0].id).toEqual('book1');
          expect(books[0].title).toEqual('Book 1');
          expect(books[0].author).toEqual('Author 1');
          // test the remaining books here
          expect(Array.isArray(books)).toBe(true); // add this line
        })
        .then((response) => {
          expect(response.body).toMatchSnapshot();
        });
    });
  });

  afterAll(async () => {
    await app.close();
  });
});
