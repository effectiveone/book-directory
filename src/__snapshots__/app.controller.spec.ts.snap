// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppController (e2e) /books (GET) should return an array of books 1`] = `
[
  {
    "author": "Author 1",
    "id": "book1",
    "title": "Book 1",
  },
  {
    "author": "Author 2",
    "id": "book2",
    "title": "Book 2",
  },
  {
    "author": "Author 3",
    "id": "book3",
    "title": "Book 3",
  },
]
`;
